lib/spack/spack/solver/asp.py

```diff
@@ -3340,7 +3343,41 @@ def variant_value(self, node, name, value):
 
         self._specs[node].update_variant_validate(name, value)
 
+    def _associated_branch(self, version_string, package_class):
+        if hasattr(package_class, "git"):
+            version = vn.Version(version_string)
+            version_dict = package_class.versions.get(version, {})
```

Comment:
The version could have a git associated with it even if the package does not. In that case you'll want to look at `package_class.versions`. This isn't tested, but I'm, thinking something along these lines:

```suggestion
        version = vn.Version(version_string)
        version_dict = package_class.get(version, {})
        git = version_dict.get("git", None) or getattr(package_class, "git", None)
        if git:
```

We can check -- if `branch` is only a valid option for git versions, then we could simplify it further.

Reply:
CVS has a branch option as well. https://github.com/spack/spack/blob/b9dfae47227a5e49f430632868c623f5065695bd/lib/spack/spack/fetch_strategy.py#L983

Reply:
@becker33 are we also taking into account package attributes that could come from the `packages.yaml` config as well here?

Reply:
Yes

Reply:
@becker33 I implemented this locally. Any thoughts on how to test the version with a custom git attribute? Trying to get a test before pushing again.